### DISCUSSION TITLE
Syntax highlighting in intellij in .ideavimrc and .vim files

### DISCUSSION THREAD

**Discussion by arabshapt** [JetBrains/ideavim#713] [General]
Hi everyone, do you know how to get file type for .vimcr and *.vim files?

**Comment by citizenmatt** [comment]
The IntelliJ Platform ships with a TextMate bundle for VimL, based on [Alhadis/language-viml](https://github.com/Alhadis/language-viml), so should already support `.vimrc` and `.vim`. The slightly modified version also recognises `.ideavimrc` files. Unfortunately, it's not possible to add custom file extensions to TextMate bundles.

**Reply by arabshapt** [reply]
Thank you for your answer) How would you modify it? And how to add the modified version?

**Reply by arabshapt** [reply]
It feels strange that an ide doesn't support its own config file...

**Reply by citizenmatt** [reply]
Not sure what you mean - which config file?

**Reply by arabshapt** [reply]
By config I mean .ideavim file.

**Reply by citizenmatt** [reply]
The `.ideavimrc` file should be correctly highlighted.

**Reply by roycrippen4** [reply]
I recently went down this rabbit hole and it was a significantly deeper hole than I had anticipated. Writing textmate syntax highlighting code is extremely difficult to wrap your head around. I was able to achieve a small amount of success by finding the built-in textmate bundles that come with JetBrains products and directly modifying the vimL textmate bundle. This ended up being a pretty bad solution with results that were barely better than nothing. I had more luck by finding other textmate bundles for vimscript and straight up replacing the contents of the built-in bundle with the ones I had found. That worked significantly better, but would occasionally cause the IDE throw warnings. I ended up giving up entirely because I had sunk probably 20 hours over a weekend trying to figure it out and struggled massively. IMO the best solution to this problem is to integrate treesitter in JetBrains products, either directly in the editors themselves, or via a plugin that can override the default syntax highlighting.

**Reply by roycrippen4** [reply]
I should add, if it wasn't already obvious, my ideavimrc file does not have syntax highlighting of any kind at the moment. The lack of syntax highlighting is not due to me messing with it either. I replaced all of my changes with a fresh copy of the vimL bundle from RustRover.

**Reply by arabshapt** [reply]
Why can I do "gcc" in actual vim and toggle comment in ideavimrc but not in intellij? It should be easy to jump in, adjust your config and reload. I hope we are not the only ones with this problem

**Reply by roycrippen4** [reply]
Wait, you can use gcc to comment while editing your .ideavimrc(within intellij), but when you edit a different file than the .ideavimrc you can't use gcc?

**Reply by arabshapt** [reply]
No,  I can't use gcc in intellij in ideavimrc. Only using neovim.

**Reply by arabshapt** [reply]
With "actual vim" I meant neovim.

**Reply by citizenmatt** [reply]
The syntax highlighting for `.ideavimrc` is just TextMate syntax highlighting - it's not full language support. As such, `gcc` doesn't know what characters to use to create comments, because it uses the IDE's own comment/uncomment actions. If IntelliJ doesn't know how to do it, IdeaVim can't do it.

@roycrippen4 syntax highlighting should just work. What version of which IDE are you using? What OS?

**Reply by roycrippen4** [reply]
@citizenmatt I've gotten only comment, string, and constant highlighting in every combination of the following. 

OS: 
Windows 10
Ubuntu 22.04

IDEs:
Webstorm
RustRover
CLion
Intellij
Ryder

**Reply by citizenmatt** [reply]
That might be related to the editor colour scheme. Here's my `~/.ideavimrc` in IDEA, with the default Light theme. It looks like only keywords, strings, comments and numbers are highlighted.

<img width="840" alt="Screenshot 2023-11-07 at 4 11 14 pm" src="https://github.com/JetBrains/ideavim/assets/222659/a5548a69-8b59-46d8-b360-f91ea81ebc0a">

The same file in Rider, with the default Rider Light theme is more colourful, with variables and functions highlighted:

<img width="840" alt="Screenshot 2023-11-07 at 4 13 48 pm" src="https://github.com/JetBrains/ideavim/assets/222659/ab36ccde-597d-4f9e-9ae7-cf7fe4f0dd0e">

Does this look similar to what you see?

**Reply by hoshsadiq** [reply]
@citizenmatt sorry for reviving an old thread, but I do not see any syntax highlighting. vim as a file type is not even in the configuration.

**Reply by citizenmatt** [reply]
Do you have TextMate support enabled? Check for the TextMate page in the settings dialog. Away from keyboard right now, but you should see Vim in that list

**Reply by hoshsadiq** [reply]
Thanks! Yup, it is enabled, and `viml` is in that list.

**Reply by citizenmatt** [reply]
And you still don't see syntax highlighting? What file are you looking at?

**Reply by hoshsadiq** [reply]
Aah okay. So it's because of the filename. Is there a way to associate a random filename (specifically `ideavimrc` without the `.` prefix) with this syntax highlight? `:set filetype=vim` does not work, nor adding `" vim: filetype=vim` to the file.

**Reply by citizenmatt** [reply]
I don't think there is a way, sorry. IIRC, the TextMate bundle has a couple of file extensions - mainly `.vim` and some known files like `.vimrc`, `.ideavimrc` and I think `_ideavimrc`. But I can't remember if there's a way to associate an arbitrary file extension with a specific TextMate bundle.

The `'filetype'` option isn't supported. I'd love to see it, but wouldn't know how to support it.

**Reply by hoshsadiq** [reply]
Aah okay. I guess I can work with this for now, or I'll simply use vim when editing this file. Thanks!

**Reply by citizenmatt** [reply]
What file are you trying to treat as Vimscript? Cant you rename it to have a `.vim` extension?
